protocol/contracts/v2/strategies/SkyFarmBaseStrategy.sol

```diff
@@ -0,0 +1,220 @@
+        daiAmount = _vaultRedemption(usdsVaultToken.maxRedeem(address(this)));
+
+        // Repay to TRV ensuring that funds stop in the TRV, they don't get pushed 
+        // back to the base strategy (ie back here)
+        treasuryReservesVault.repay(daiToken, daiAmount, address(this));
+    }
+
+    /**
+     * @notice The TRV sends the tokens to deposit (and also mints equivalent dTokens)
+     * The strategy is then expected to put those tokens to work.
+     */
+    function trvDeposit(uint256 amount) external override {
+        if (msg.sender != address(treasuryReservesVault)) revert OnlyTreasuryReserveVault(msg.sender);
+        if (amount == 0) revert CommonEventsAndErrors.ExpectedNonZero();
+        _daiDeposit(amount);
+    }
+
+    /**
+     * @notice The TRV is able to withdraw on demand in order to fund other strategies which 
+     * wish to borrow from the TRV.
+     * @dev It may withdraw less than requested if there isn't enough balance in the `usdsVaultToken`.
+     */
+    function trvWithdraw(uint256 requestedAmount) external override returns (uint256) {
+        address _trvAddr = address(treasuryReservesVault);
+        if (msg.sender != _trvAddr) revert OnlyTreasuryReserveVault(msg.sender);
+        if (requestedAmount == 0) revert CommonEventsAndErrors.ExpectedNonZero();       
+
+        // Cap at the max amount which can be withdrawn.
+        uint256 availableToWithdraw = usdsVaultToken.maxWithdraw(address(this));
+        if (requestedAmount > availableToWithdraw) {
+            requestedAmount = availableToWithdraw;
+        }
+
+        _daiWithdrawal(requestedAmount);
+        daiToken.safeTransfer(_trvAddr, requestedAmount);
+        return requestedAmount;
+    }
+
+    /**
+     * @notice A hook where strategies can optionally update approvals when the trv is updated
+     */
+    function _updateTrvApprovals(address oldTrv, address newTrv) internal override {
+        _setMaxAllowance(daiToken, oldTrv, newTrv);
+    }
+
+    function _daiDeposit(uint256 daiAmount) internal {
+        // Convert to USDS
+        daiToUsds.daiToUsds(address(this), daiAmount);
+
+        // Deposit into the USDS savings vault
+        usdsVaultToken.deposit(daiAmount, address(this));
+        emit DaiDeposited(daiAmount);
+    }
+
+    function _daiWithdrawal(uint256 daiAmount) internal {
+        // Withdraw from the USDS savings vault
+        usdsVaultToken.withdraw(daiAmount, address(this), address(this));
+
+        // Convert to DAI
+        daiToUsds.usdsToDai(address(this), daiAmount);
+        emit DaiWithdrawn(daiAmount);
+    }
+
+    function _vaultRedemption(uint256 vaultShares) internal returns (uint256 daiAmount) {
+        // Redeem from the USDS savings vault
+        daiAmount = usdsVaultToken.redeem(vaultShares, address(this), address(this));
+
+        // Convert to DAI
+        daiToUsds.usdsToDai(address(this), daiAmount);
```

Comment:
nit: 
As this function is used as part of the shutdown, you could consider converting and depositing all the DAI balance of the contract (accounting for donations). But not a big deal, since you can later call the `recoverToken()` function. So perhaps it is better to keep it clean here.